sandbox/cgroup/monitor.go

```diff
@@ -0,0 +1,189 @@
+	// incremented by one (of course, if a parent path isn't in the map,
+	// it is initialized to 'one'). Every time the path of a CGroup is deleted,
+	// the corresponding usage counter here corresponding to the parent path
+	// is decremented, and when it reaches zero, it is removed and inotify is
+	// notified to stop monitoring it.
+	pathList map[string]int32
+}
+
+// Contains the data corresponding to a CGroup that must be watched to detect
+// when it is destroyed
+type groupToWatch struct {
+	// This is the CGroup identifier
+	name string
+	// This contains a list of folders to monitor. When all of them have been
+	// deleted, the CGroup has been destroyed and there are no processes running
+	folders []string
+	// This channel is used to notify to the requester that this CGroup has been
+	// destroyed. The watcher writes the CGroup identifier on it; this way, the
+	// same channel can be shared to monitor several CGroups.
+	channel chan<- string
+}
+
+var currentWatcher *inotifyWatcher = &inotifyWatcher{
+	wd:           nil,
+	pathList:     make(map[string]int32),
+	addWatchChan: make(chan *groupToWatch),
+}
+
+func (iw *inotifyWatcher) addWatch(newWatch *groupToWatch) {
+	var folderList []string
+	for _, fullPath := range newWatch.folders {
+		// It's not possible to use inotify.InDeleteSelf in /sys/fs because it
+		// isn't triggered, so we must monitor the parent folder and use InDelete
+		basePath := path.Dir(fullPath)
+		if _, exists := iw.pathList[basePath]; !exists {
+			iw.pathList[basePath] = 0
+			if err := iw.wd.AddWatch(basePath, inotify.InDelete); err != nil {
+				delete(iw.pathList, basePath)
+				continue
+			}
+		}
+		iw.pathList[fullPath]++
+		if osutil.FileExists(fullPath) {
+			folderList = append(folderList, fullPath)
+		} else {
+			iw.removePath(fullPath)
+		}
+	}
+	if len(folderList) == 0 {
+		newWatch.channel <- newWatch.name
+	} else {
+		newWatch.folders = folderList
+		iw.groupList = append(iw.groupList, newWatch)
+	}
+}
+
+func (iw *inotifyWatcher) removePath(fullPath string) {
+	iw.pathList[fullPath]--
+	if iw.pathList[fullPath] == 0 {
+		iw.wd.RemoveWatch(fullPath)
+		delete(iw.pathList, fullPath)
+	}
+}
+
+// processDeletedPath checks if the received "path corresponds to the passed
+// CGroup, removing it from the list of folders being watched in that CGroup if
+// needed. It returns TRUE if there remain folders to be monitored in that CGroup,
+// or FALSE if all the folders of that CGroup have been deleted.
+func (iw *inotifyWatcher) processDeletedPath(watch *groupToWatch, deletedPath string) bool {
+	for i, fullPath := range watch.folders {
+		if fullPath == deletedPath {
+			// if the folder name is in the list of folders to monitor, decrement the
+			// parent's usage counter, and remove it from the list of folders to watch
+			// in this CGroup (by not adding it to tmpFolders)
+			iw.removePath(fullPath)
+			watch.folders = append(watch.folders[:i], watch.folders[i+1:]...)
```

Comment:
Is this missing a `break` or can it have the same path more than once? If it can, than the re-slicing is a problem

Reply:
You are right: the path guaranteed to be only once, so a break is a good idea. Done.